[interpreter] Fix compilation warnings; upgrade to OCaml >=4.07

mdesharnais opened the pull request:
Fix deprecated warnings for String.uppercase (deprecated since 4.03.0) and Pervasives (deprecated since 4.08.0). Breaks compatibility with OCaml < 4.07.0. If compatibility is required, [stdlib-shims](https://github.com/ocaml/stdlib-shims) could be used.

mdesharnais commented:
I think the simplest is to refer to the ocaml.org documentation on installing OCaml. The people there seem to make a good job at tracking changes on different platforms.

rossberg commented:
Yeah, except for Windows. :( The second Windows link points to an installer that hasn't been updated since 4.01, the fourth now points to some odd crypto trading page. OPAM install isn't even available on Windows.

mdesharnais commented:
You are right, I actually just checked the first link ([Windows OPAM repository](https://fdopen.github.io/opam-repository-mingw/)) and this one is up to date with OCaml 4.10. I found an [issue](https://github.com/ocaml/ocaml.org/issues/1097) (dated 7 Dec. 2019) on the website's tracker concerning the broken link (Bash on Windows), but nobody fixed it.

I just submitted a [pull request](https://github.com/ocaml/ocaml.org/pull/1123) fixing the issues. If it gets merge, do you think we may not give any specific Windows instruction in the README?

rossberg commented:
Let's try and see whether somebody complains. Be prepared to roll it back in that case, though. ;)

As for the README, we probably still want to keep the winmake option around for now, in case somebody has a bare OCaml build running outside Cygwin.

Would you mind also incorporating PR #1109 into this one? Then a rollback is easier.

dulanov commented:
May be it can be simpler to use Nix, not only by Win10/WSL as I use it, but also on Linux.

First of all you need something like `ocaml.nix`:

```
with import <nixpkgs> {};
stdenv.mkDerivation {
  name = "ocaml-env";

  buildInputs = [
    ocaml
    ocamlbuild
  ];
}
```

And apply it in you env like: `nix-env ocaml.nix`. After that `make` works also on Win10.

Here is warnings dump from branch - w3c-1.0:

```
[nix-shell:/mnt/d/References/spec/interpreter]$ make
echo >_tags "true: bin_annot"
echo >>_tags "true: debug"
echo >>_tags "<{util,syntax,binary,text,valid,runtime,exec,script,host,main}/*.cmx>: for-pack(Wasm)"
ocamlbuild -cflags '-w +a-4-27-42-44-45 -warn-error +a-3' -I util -I syntax -I binary -I text -I valid -I runtime -I exec -I script -I host -I main -libs bigarray -quiet main.native
+ ocamlc.opt -c -w +a-4-27-42-44-45 -warn-error +a-3 -g -bin-annot -I exec -I valid -I util -I text -I syntax -I script -I runtime -I main -I host -I binary -o exec/float.cmo exec/float.ml
File "exec/float.ml", line 140, characters 22-37:
140 |   let sqrt  x = unary Pervasives.sqrt x
                            ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 142, characters 22-37:
142 |   let ceil  x = unary Pervasives.ceil x
                            ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 143, characters 22-38:
143 |   let floor x = unary Pervasives.floor x
                            ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 150, characters 29-44:
150 |     let f = if xf < 0.0 then Pervasives.ceil xf else Pervasives.floor xf in
                                   ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 150, characters 53-69:
150 |     let f = if xf < 0.0 then Pervasives.ceil xf else Pervasives.floor xf in
                                                           ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 159, characters 12-27:
159 |     let u = Pervasives.ceil xf in
                  ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 160, characters 12-28:
160 |     let d = Pervasives.floor xf in
                  ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 165, characters 36-52:
165 |       um = dm && let h = u /. 2. in Pervasives.floor h = h
                                          ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 234, characters 14-30:
234 |     let s1' = String.uppercase s1 in
                    ^^^^^^^^^^^^^^^^
Alert deprecated: Stdlib.String.uppercase
Use String.uppercase_ascii instead.
File "exec/float.ml", line 235, characters 14-30:
235 |     let s2' = String.uppercase s2 in
                    ^^^^^^^^^^^^^^^^
Alert deprecated: Stdlib.String.uppercase
Use String.uppercase_ascii instead.
File "exec/float.ml", line 269, characters 27-43:
269 |       let i = skip_zeroes (String.uppercase s) 0 in
                                 ^^^^^^^^^^^^^^^^
Alert deprecated: Stdlib.String.uppercase
Use String.uppercase_ascii instead.
+ ocamlopt.opt -c -w +a-4-27-42-44-45 -warn-error +a-3 -g -bin-annot -for-pack Wasm -I exec -I valid -I util -I text -I syntax -I script -I runtime -I main -I host -I binary -o exec/float.cmx exec/float.ml
File "exec/float.ml", line 140, characters 22-37:
140 |   let sqrt  x = unary Pervasives.sqrt x
                            ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 142, characters 22-37:
142 |   let ceil  x = unary Pervasives.ceil x
                            ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 143, characters 22-38:
143 |   let floor x = unary Pervasives.floor x
                            ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 150, characters 29-44:
150 |     let f = if xf < 0.0 then Pervasives.ceil xf else Pervasives.floor xf in
                                   ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 150, characters 53-69:
150 |     let f = if xf < 0.0 then Pervasives.ceil xf else Pervasives.floor xf in
                                                           ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 159, characters 12-27:
159 |     let u = Pervasives.ceil xf in
                  ^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 160, characters 12-28:
160 |     let d = Pervasives.floor xf in
                  ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 165, characters 36-52:
165 |       um = dm && let h = u /. 2. in Pervasives.floor h = h
                                          ^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the
stdlib-shims library: https://github.com/ocaml/stdlib-shims
File "exec/float.ml", line 234, characters 14-30:
234 |     let s1' = String.uppercase s1 in
                    ^^^^^^^^^^^^^^^^
Alert deprecated: Stdlib.String.uppercase
Use String.uppercase_ascii instead.
File "exec/float.ml", line 235, characters 14-30:
235 |     let s2' = String.uppercase s2 in
                    ^^^^^^^^^^^^^^^^
Alert deprecated: Stdlib.String.uppercase
Use String.uppercase_ascii instead.
File "exec/float.ml", line 269, characters 27-43:
269 |       let i = skip_zeroes (String.uppercase s) 0 in
                                 ^^^^^^^^^^^^^^^^
Alert deprecated: Stdlib.String.uppercase
Use String.uppercase_ascii instead.
mv main.native wasm
rm _tags
```

mdesharnais commented:
The [pull request](https://github.com/ocaml/ocaml.org/pull/1123) to update Windows instructions was merged and is [now online](https://ocaml.org/docs/install.html#Windows).

I added the three commits from PR #1109 but squashed them and changed the commit message.

I reinserted the build instructions for Windows and try to adapt them a slightly to the installation instruction. Feel free to improve on that.

I don't know why, but after I rebased the branch on upsteam/master, the commit 4934c66 shows up in the pull request even though it is already on master. Any idea of what is happening?

@dulanov That is an interesting suggestion, but I would argue that it deserves its own pull request. With this one I really just wanted to get rid of the deprecation warning and I already think that there are many changes in there. I usually prefer to keep things separated.

rossberg commented on interpreter/README.md:
```diff
@@ -15,7 +15,7 @@ The text format defines modules in S-expression syntax. Moreover, it is generali
 ## Building
 
-You'll need OCaml 4.02 or higher. An easy way to get this on Linux is to download the [source tarball from our mirror of the ocaml website](https://wasm.storage.googleapis.com/ocaml-4.02.2.tar.gz) and do the configure / make dance.  On macOS, with [Homebrew](http://brew.sh/) installed, simply `brew install ocaml ocamlbuild`.
+You'll need OCaml 4.07 or higher. Instructions for installing a recent version of OCaml on multiple platforms are available [here](https://ocaml.org/docs/install.html). The recommended way is through [OPAM](https://ocaml.org/docs/install.html#OPAM).
```
```suggestion
You'll need OCaml 4.07 or higher. Instructions for installing a recent version of OCaml on multiple platforms are available [here](https://ocaml.org/docs/install.html). On most platforms, the recommended way is through [OPAM](https://ocaml.org/docs/install.html#OPAM).
```

rossberg commented on interpreter/README.md:
```diff
@@ -43,17 +43,21 @@ That builds `all`, plus updates `winmake.bat`.
 #### Building on Windows
 
-We recommend a pre-built installer. With [this one](https://protz.github.io/ocaml-installer/) you have two options:
+The instructions depend on how you installed OCaml.
```
```suggestion
The instructions depend on how you [installed OCaml on Windows](https://ocaml.org/docs/install.html#Windows).
```

rossberg commented on interpreter/README.md:
```diff
@@ -43,17 +43,21 @@ That builds `all`, plus updates `winmake.bat`.
+The instructions depend on how you installed OCaml.
 
-1. Bare OCaml. If you just want to build the interpreter and don't care about modifying it, you don't need to install the Cygwin core that comes with the installer. Just install OCaml itself and run
+1. *WSL*: You can build the interpreter using `make`, as described above.
+
+2. *Cygwin*: If you want to build a native code executable, or want to hack on the interpreter (i.e., use incremental compilation), then you need to install the Cygwin core that is included with the OCaml installer. Then you can build the interpreter using `make` in the Cygwin terminal, as described above.
+
```
```suggestion
1. *Cygwin*: If you want to build a native code executable, or want to hack on the interpreter (i.e., use incremental compilation), then you need to install the Cygwin core that is included with the OCaml installer. Then you can build the interpreter using `make` in the Cygwin terminal, as described above.

2. *Windows Subsystem for Linux* (WSL): You can build the interpreter using `make`, as described above.

```

rossberg commented on interpreter/README.md:
```diff
@@ -43,17 +43,21 @@ That builds `all`, plus updates `winmake.bat`.
 in a Windows shell, which creates a program named `wasm`. Note that this will be a byte code executable only, i.e., somewhat slower.
 
-2. OCaml + Cygwin. If you want to build a native code executable, or want to hack on the interpreter (i.e., use incremental compilation), then you need to install the Cygwin core that is included with the OCaml installer. Then you can build the interpreter using `make` in the Cygwin terminal, as described above.
+In any way, in order to run the test suite you'll need to have Python installed. If you used Option 2, you can invoke the test runner `runtests.py` directly instead of doing it through `make`.
```
```suggestion
In any way, in order to run the test suite you'll need to have Python installed. If you used Option 3, you can invoke the test runner `runtests.py` directly instead of doing it through `make`.
```

rossberg commented on interpreter/README.md:
```diff
@@ -43,17 +43,21 @@ That builds `all`, plus updates `winmake.bat`.
+In any way, in order to run the test suite you'll need to have Python installed. If you used Option 2, you can invoke the test runner `runtests.py` directly instead of doing it through `make`.
 
-Either way, in order to run the test suite you'll need to have Python installed. If you used Option 1, you can invoke the test runner `runtests.py` directly instead of doing it through `make`.
+The main developpement is done on Linux systems so, please, share your experience building on Windows with us!
```
```suggestion
```

mdesharnais commented:
Looks good to me. Do you want to merge with the extra code review commit or do you want me to squash it into the README commit?

rossberg commented:
You'll ultimately have to squash every PR anyway.

mdesharnais commented:
> You'll ultimately have to squash every PR anyway.

All commits into one? Don't you want to keep them separated?

rossberg commented:
Yes, this repo allows "Squash and merge" only.

mdesharnais commented:
Ok, do I have anything else to do?

Ms2ger commented:
"Rebase" works as well, but it's probably better to squash the commits in this case. However, you'll first need to fix:

```
File "exec/float.ml", line 141, characters 22-33:

Error: Unbound module Stdlib
```

mdesharnais commented:
Ok, I squashed all commits into one and changed the travis configuration to install OCaml 4.07.01 through OPAM. This should gives more flexibility to change versions of OCaml and tooling (e.g. ocamlbuild) in the future.

rossberg commented:
Oh, sorry, you didn't need to squash manually. What I meant was that GitHub will do that anyway when I press the merge button. ;)

mdesharnais commented:
> Oh, sorry, you didn't need to squash manually. What I meant was that GitHub will do that anyway when I press the merge button. ;)

No problem, now I know for next time. :)
Thanks for merging
